hellodagger/src/main/java/com/airbnb/mvrx/helloDagger/di/AssistedViewModelFactory.kt

```diff
@@ -0,0 +1,13 @@
+package com.airbnb.mvrx.helloDagger.di
+
+import com.airbnb.mvrx.MvRxState
+import com.airbnb.mvrx.helloDagger.base.MvRxViewModel
+
+/**
+ * Creates and returns instances of the requested ViewModel.
+ *
+ * This interface serves as a supertype for AssistedInject factories in ViewModels.
```

Comment:
Can you add a code snippet? It's not clear what this does with somebody who has no context.

Reply:
How does this look?
```
/**
 * Creates and returns instances of the requested ViewModel.
 *
 * This interface serves as a supertype for AssistedInject factories in ViewModels.
 * We can use this as a marker interface in a Multibinding Dagger setup to populate a map
 * of ViewModel classes with their AssistedInject factories.
 *
 * Here's an example for such a setup:
 *
 * class MyViewModel @AssistedInject constructor(...): BaseMvRxViewModel<MyState>(...) {
 *   @AssistedInject.Factory
 *   interface Factory : AssistedViewModelFactory<MyViewModel, MyState> {
 *     override fun create(state: MyState): MyViewModel
 *   }
 * }
 *
 * @AssistedModule(includes = [AssistedInject_MyAppModule::class])
 * @Module
 * interface MyAppModule {
 *   @Binds
 *   @IntoMap
 *   @ViewModelKey(MyViewModel::class)
 *   fun myViewModelFactory(factory: MyViewModel.Factory): AssistedViewModelFactory<*, *>
 * }
 *
 * Now we can request Dagger for an instance of Map<Class<out MvRxViewModel<*>>, AssistedViewModelFactory<*, *>>,
 * and use it to access our ViewModel Factories.
 */
interface AssistedViewModelFactory<VM: MvRxViewModel<S>, S: MvRxState> {
    fun create(state: S): VM
}
```

Reply:
Better, for this example, I think more documentation and less code makes sense.

Reply:
I've revised the documentation for this interface, and have added more explanations to all the parts. It is longer, but I think it conveys the message a lot better now:

```kotlin
/*
 * Serves as a supertype for AssistedInject factories in ViewModels.
 *
 * We can use this interface as a marker in a Multibinding Dagger setup to populate a Map
 * of ViewModel classes with their AssistedInject factories.
 *
 * This setup is needed because AssistedInject factories are isolated from each other, so this
 * interface provides a common parent type to facilitate grouping these factories into a collection.
 * When we use it correctly with Dagger, a Map of ViewModels and their factories can be created for us
 * automatically. This Map can then be used to retrieve for a ViewModel's factory using its class.
 *
 * Here's an example for such a setup:
 *
 * First we define our ViewModel with an @AssistedInject annotated constructor, and a Factory interface
 * implementing AssistedViewModelFactory.
 *
 * class MyViewModel @AssistedInject constructor(...): BaseMvRxViewModel<MyState>(...) {
 *   @AssistedInject.Factory
 *   interface Factory : AssistedViewModelFactory<MyViewModel, MyState> {
 *     override fun create(state: MyState): MyViewModel
 *   }
 * }
 *
 * Then we need to create a Dagger Module, which contains methods that @Binds @IntoMap all of our
 * AssistedViewModelFactories using a [ViewModelKey]. Notice that the input to these methods is
 * the exact type of our AssistedInject factory, but the return type is an AssistedViewModelFactory.
 *
 * @AssistedModule(includes = [AssistedInject_MyAppModule::class])
 * @Module
 * interface MyAppModule {
 *   @Binds
 *   @IntoMap
 *   @ViewModelKey(MyViewModel::class)
 *   fun myViewModelFactory(factory: MyViewModel.Factory): AssistedViewModelFactory<*, *>
 * }
 *
 * This Module tells Dagger to include MyViewModel.Factory class in the Multibinding map using
 * MyViewModel::class as the key. Such a method should be added for **every ViewModel Factory**
 * so that they can be identified by Dagger and used for populating the Map.
 *
 * The generated map can then be injected wherever it is required.
 *
 * class BaseActivity: BaseMvRxActivity() {
 *   @Inject
 *   lateinit var factories: @JvmSuppressWildcards Map<Class<out BaseViewModel<*>>, AssistedViewModelFactory<*, *>>
 *
 *   onCreate(...) {
 *     appComponent().inject(this)
 *   }
 * }
 */
interface AssistedViewModelFactory<VM: BaseViewModel<S>, S: MvRxState> {
    fun create(state: S): VM
}
```